packages/transaction-controller/src/TransactionController.ts

```diff
@@ -1695,6 +1730,164 @@ export class TransactionController extends BaseController<
       currentNetworkIsEIP1559Compatible && currentAccountIsEIP1559Compatible
     );
   }
+
+  private async updatePostTransactionBalance(
+    transactionMeta: TransactionMeta,
+    { numberOfAttempts = 6 },
+  ) {
+    const transactionId = transactionMeta.id;
+    const postTransactionBalance = await query(this.ethQuery, 'getBalance', [
+      transactionMeta.txParams.from,
+    ]);
+    const latestTransactionMeta = this.getTransaction(
+      transactionId,
+    ) as TransactionMeta;
+    const approvalTransactionMeta = latestTransactionMeta.approvalTxId
+      ? this.getTransaction(latestTransactionMeta.approvalTxId)
+      : null;
+    latestTransactionMeta.postTxBalance = postTransactionBalance.toString(16);
+    const isDefaultTokenAddress = isSwapsDefaultTokenAddress(
+      transactionMeta.destinationTokenAddress as string,
+      transactionMeta.chainId,
+    );
+
+    if (
+      isDefaultTokenAddress &&
+      transactionMeta.preTxBalance === latestTransactionMeta.postTxBalance &&
+      numberOfAttempts > 0
+    ) {
+      setTimeout(() => {
```

Comment:
Since this is already an `async` method, would it be cleaner to refactor this recursion into a simple `for` loop using a basic `sleep` method that we `await` rather using these callbacks.

It would also mean that this method resolves when the actual balance is updated which is more explicit but we'd have to remember not to `await` when we call it.

Reply:
This is done, had to disable some lint errors in order to make tests work. https://github.com/MetaMask/core/pull/1877/commits/127444c3b779230c85656050ea48613c3ff3de71